### DISCUSSION TITLE
Java Buildpacks projects

### DISCUSSION THREAD

**Discussion by cmoulliard** [buildpacks/community#119] [Ideas]
## Hosting of Java Buildpacks projects

Our Red Hat snowdrop team has designed/developed different java projects that we would like to share with the buildpacks community : 

- Java Buildpack client supporting to build a project using a container running the lifecycle creator - https://github.com/snowdrop/java-buildpack-client
- Java Buildpack dummy project - https://github.com/snowdrop/snowdrop-buildpacks#development-of-a-buildpacks-using-quarkus-and-native-build able to generate a native executable `main` supporting `detect` or `build`. Soon, this project will be comparable to what it has been designed by `packit` or `libcnb`
  
  **Remark**: This project should be extracted from the `snowdrop-buildpacks` and repacked as `SDK/lib` and `API` supporting the Model of the `Buildpack API`

**Questions**

- Is the buildpacks community interested to contribute to the development of such java projects ?
- If yes, do you agree that we host the projects under `github.com/buildpacks` org ?

**Comment by jromero** [comment]
The Java Buildpack Client might actually help us with a current [Jenkins plugin](https://github.com/fatiiates/buildpacks-jenkins-plugin) integration.

cc: @fatiiates

**Comment by dmikusa** [comment]
I strongly disagree that we should host language-specific buildpacks under buildpacks.io. That is going to create a defacto standard and cripple any community efforts at having similar buildpacks. 

Right now there are no buildpacks under buildpacks.io, it's just the spec (which I personally like). There is a [PR to make some utility buildpacks officially part of buildpacks.io](https://github.com/buildpacks/rfcs/pull/175), so there's a sentiment that we need some buildpacks. This PR is suggesting a pretty limited set of buildpacks meeting a [certain criteria](https://github.com/buildpacks/rfcs/blob/a735d9f233b802391b16e43627f17da07404183d/text/0000-official-utility-buildpacks.md#how-it-works), I don't think the intention is to take language-specific buildpacks at this time.

Anyway, if you are looking for ways to share and promote your buildpacks through the community I think you want the [Buildpack Registry](https://registry.buildpacks.io/). Anyone can register a buildpack there and share it with others. There's also a search UI so folks can find interesting and helpful buildpacks.

**Reply by jromero** [reply]
I think there may be a misunderstanding here. The way I understand the original post is to provide a set of java buildpack **libraries**. It would be similar alternatives to `pack`, used as a go library, as a **Client** and libcnb as a go buildpack **SDK**.

**Reply by dmikusa** [reply]
+1 for libcnb in Java
-1 for actual buildpacks that provide Java

The topic of more libraries like libcnb came up on Slack and two questions shook out.

1. What requirements do the Buildpacks project dictate for a library to be considered?
2. What can we do as a project to help interested parties grow & develop their libraries and get them ready to meet the criteria for #1. 

IMHO, those need to be worked out before we accept any additional libraries.

**Comment by cmoulliard** [comment]
FYI. The purpose of the project `Java Buildpack client` has nothing todo with the idea to offer Java buildpacks but instead java classes supporting : 
- An implementation of the buildpack platform spec that can be used to build projects using a specified buildpacks.
- A fluent interface to allow creation & configuration of the build to be performed.
...

The client uses the combined `creator lifecycle` phase from the buildpack to drive the entire build. 

Example
```
BuildpackBuilder.get()
    .withContent(new File("/home/user/java-project"))
    .withFinalImage("test/testimage:latest")
    .build();
```

**Remark**: This project could be extended to include, like the project `libcnb`, additional classes able to by example do a build, detect as implemented as dummy examples here: https://github.com/snowdrop/snowdrop-buildpacks/blob/main/buildpacks/java/jdk/src/main/java/dev/snowdrop/buildpack/Build.java or https://github.com/snowdrop/snowdrop-buildpacks/blob/main/buildpacks/java/jdk/src/main/java/dev/snowdrop/buildpack/Detect.java
